File: Task/Chebyshev-coefficients/Zen-C/chebyshev-coefficients.c
import "std/math.zc";
import "std/string.zc";

alias cheb_fn = fn*(f64) -> f64;

fn map_range(x: f64, min: f64, max: f64, min_to: f64, max_to: f64) -> f64 {
    return (x - min) / (max - min) * (max_to - min_to) + min_to;
}

fn cheb_coeffs(func: cheb_fn, n: int, min: f64, max: f64, coeffs: f64*) {
    for i in 0..n {
        let ii = (f64)i;
        let nn = (f64)n;
        let f = func(map_range(Math::cos(Math::PI() * (ii + 0.5) / nn), -1.0, 1.0, min, max));
        f = f * 2.0 / nn;
        for j in 0..n {
            let jj = (f64)j;
            coeffs[j] += f * Math::cos(Math::PI() * jj * (ii + 0.5) / nn);
        }
    }
}

fn cheb_approx(x: f64, n: int, min: f64, max: f64, coeffs: f64*) -> f64 {
    assert(n >= 2, "'n' can't be less than 2.";
    let a = 1.0;
    let b = map_range(x, min, max, -1.0, 1.0);
    let res = coeffs[0] / 2.0 + coeffs[1] * b;
    let xx = 2.0 * b;
    for i in 2..n {
        let c = xx * b - a;
        res += coeffs[i] * c;
        a = b;
        b = c;
    }
    return res;
}

fn cheb_func(x: f64)-> f64 { return Math::cos(x); }

fn main() {
    def N = 10;
    let min = 0.0;
    let max = 1.0;
    let coeffs: [f64; N];
    cheb_coeffs(cheb_func, N, min, max, coeffs);
    println "Coefficients:";
    for coeff in coeffs {
        let s = coeff >= 0 ? " ": "";
        println "{s}{coeff:1.14f}";
    }
    println "\nApproximations:\n  x      func(x)    approx       diff";
    for i in 0..=20 {
        let x = map_range((f64)i, 0.0, 20.0, min, max);
        let f = Math::cos(x);
        let approx = cheb_approx(x, N, min, max, coeffs);
        let diff = approx - f;
        let ds = String::from("{diff:g}");
        let len = ds.length();
        let e = ds.substring(len - 4, 4);
        let ds2 = ds.substring(0, len - 5);
        let ds3: String;
        if diff >= 0 {
            ds3 = ds2.substring(0, 4);
            ds3.insert_rune(0, ' ');
        } else {
            ds3 = ds2.substring(0, 5);
        }
        ds3.append(&e);
        println "{x:1.3f}  {f:1.8f} {approx:1.8f}  {ds3}";
    }
}

File: Task/Chebyshev-coefficients/Swift/chebyshev-coefficients.swift
import Foundation

typealias DFunc = (Double) -> Double

func mapRange(x: Double, min: Double, max: Double, minTo: Double, maxTo: Double) -> Double {
  return (x - min) / (max - min) * (maxTo - minTo) + minTo
}

func chebCoeffs(fun: DFunc, n: Int, min: Double, max: Double) -> [Double] {
  var res = [Double](repeating: 0, count: n)

  for i in 0..<n {
    let dI = Double(i)
    let dN = Double(n)
    let f = fun(mapRange(x: cos(.pi * (dI + 0.5) / dN), min: -1, max: 1, minTo: min, maxTo: max)) * 2.0 / dN

    for j in 0..<n {
      res[j] += f * cos(.pi * Double(j) * (dI + 0.5) / dN)
    }
  }

  return res
}

func chebApprox(x: Double, n: Int, min: Double, max: Double, coeffs: [Double]) -> Double {
  var a = 1.0
  var b = mapRange(x: x, min: min, max: max, minTo: -1, maxTo: 1)
  var res = coeffs[0] / 2.0 + coeffs[1] * b
  let xx = 2 * b
  var i = 2

  while i < n {
    let c = xx * b - a
    res += coeffs[i] * c
    (a, b) = (b, c)
    i += 1
  }

  return res
}

let coeffs = chebCoeffs(fun: cos, n: 10, min: 0, max: 1)

print("Coefficients")

for coeff in coeffs {
  print(String(format: "%+1.15g", coeff))
}

print("\nApproximations:\n  x      func(x)     approx       diff")

for i in stride(from: 0.0, through: 20, by: 1) {
  let x = mapRange(x: i, min: 0, max: 20, minTo: 0, maxTo: 1)
  let f = cos(x)
  let approx = chebApprox(x: x, n: 10, min: 0, max: 1, coeffs: coeffs)

  print(String(format: "%1.3f  %1.8f  %1.8f  % 4.1e", x, f, approx, approx - f))
}
